File: client/csharp/src/Event.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using KRPC.Schema.KRPC;

namespace KRPC.Client
{
    /// <summary>
    /// Object representing an event.
    /// </summary>
    public class Event : IEquatable<Event>
    {
        readonly Stream<bool> stream;

        internal Event (Connection connection, KRPC.Schema.KRPC.Event evnt)
        {
            stream = new Stream<bool> (connection, evnt.Stream.Id);
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public override bool Equals (object obj)
        {
            if (ReferenceEquals (this, obj))
                return true;
            if (ReferenceEquals (obj, null))
                return false;
            var typedObj = obj as Event;
            return typedObj != null && Equals (typedObj);
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public bool Equals (Event other)
        {
            return !ReferenceEquals (other, null) && stream == other.stream;
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public static bool operator == (Event lhs, Event rhs)
        {
            if (ReferenceEquals (lhs, null) || ReferenceEquals (rhs, null))
                return ReferenceEquals (lhs, rhs);
            if (ReferenceEquals (lhs, rhs))
                return true;
            return lhs.Equals (rhs);
        }

        /// <summary>
        /// Returns true if the objects are not equal.
        /// </summary>
        public static bool operator != (Event lhs, Event rhs)
        {
            if (ReferenceEquals (lhs, null) || ReferenceEquals (rhs, null))
                return !ReferenceEquals (lhs, rhs);
            if (ReferenceEquals (lhs, rhs))
                return false;
            return !(lhs.Equals (rhs));
        }

        /// <summary>
        /// Hash code for the object.
        /// </summary>
        public override int GetHashCode ()
        {
            return stream.GetHashCode ();
        }

        /// <summary>
        /// Start the event
        /// </summary>
        public void Start() {
            stream.Start(false);
        }

        /// <summary>
        /// Condition variable that is notified when the event occurs.
        /// </summary>
        public object Condition {
            get { return stream.Condition; }
        }

        /// <summary>
        /// Wait until the event occurs.
        /// </summary>
        public void Wait (double timeout = -1) {
            if (!stream.stream.Started)
                Start();
            stream.stream.Value = false;
            while (!stream.Get ()) {
                var origValue = stream.Get ();
                stream.Wait (timeout);
                if (timeout >= 0 && stream.Get () == origValue)
                    // Value did not change, must have timed out
                    return;
            }
        }

        /// <summary>
        /// Add a callback that is invoked whenever the stream is updated.
        /// </summary>
        public void AddCallback (Action callback) {
            stream.AddCallback ((bool x) => { if (x) callback(); });
        }

        /// <summary>
        /// The underlyling stream for the event.
        /// </summary>
        public Stream<bool> Stream {
            get { return stream; }
        }

        /// <summary>
        /// Remove the event from the server.
        /// </summary>
        public void Remove ()
        {
            stream.Remove ();
        }
    }
}

File: client/cpp/src/event.cpp
#include "krpc/event.hpp"

#include <memory>
#include <string>

#include "krpc/krpc.pb.hpp"
#include "krpc/stream_impl.hpp"

namespace krpc {

class Client;

Event::Event() {
}

Event::Event(Client * client, const krpc::schema::Event& message) :
  _stream(client, message.stream().id()) {
}

void Event::start() {
  _stream.start(false);
}

std::condition_variable& Event::get_condition() const {
  return _stream.get_condition();
}

std::unique_lock<std::mutex>& Event::get_condition_lock() const {
  return _stream.get_condition_lock();
}

void Event::acquire() {
  _stream.acquire();
}

void Event::release() {
  _stream.release();
}

void Event::wait(double timeout) {
  start();
  _stream.impl->update(std::string("\x00", 1), nullptr);
  while (!_stream()) {
    bool origValue = _stream();
    _stream.wait(timeout);
    if (timeout >= 0 && _stream() == origValue)
      // Value did not change, must have timed out
      return;
  }
}

int Event::add_callback(const Callback& callback) {
  return _stream.add_callback([callback] (bool _) { callback(); });
}

void Event::remove_callback(int tag) {
  return _stream.remove_callback(tag);
}

void Event::remove() {
  _stream.remove();
}

Stream<bool> Event::stream() const {
  return _stream;
}

bool Event::operator==(const Event& rhs) const {
  return _stream == rhs._stream;
}

bool Event::operator!=(const Event& rhs) const {
  return _stream != rhs._stream;
}

Event::operator bool() const {
  return _stream.operator bool();
}

}  // namespace krpc
